ecosystem/indexer-grpc/indexer-grpc-file-store/src/processor.rs

```diff
@@ -87,109 +84,174 @@ impl Processor {
-            let process_size = transactions_buffer.len() / BLOB_STORAGE_SIZE * BLOB_STORAGE_SIZE;
-            let current_batch: Vec<EncodedTransactionWithVersion> =
-                transactions_buffer.drain(..process_size).collect();
-            let last_transaction = current_batch.as_slice().last().unwrap().clone();
-            self.file_store_operator
-                .upload_transactions(cache_chain_id, current_batch)
+            // update next batch start version
+            batch_start_version = last_version + 1;
+            assert!(
+                batch_start_version % BLOB_STORAGE_SIZE as u64 == 0,
+                "[Filestore] Batch must be multiple of 1000"
+            );
+
+            // Update filestore metadata. First do it in cache for performance then update metadata file
+            let start_time = std::time::Instant::now();
+            self.cache_operator
+                .update_file_store_latest_version(batch_start_version)
+                .await?;
+            while self
+                .file_store_operator
+                .update_file_store_metadata_with_timeout(cache_chain_id, batch_start_version)
                 .await
-                .context("Uploading transactions to file store failed.")?;
-            PROCESSED_VERSIONS_COUNT.inc_by(process_size as u64);
-            tps_calculator.tick_now(process_size as u64);
-            let end_version = current_file_store_version + process_size as u64 - 1_u64;
-            let num_transactions = end_version - current_file_store_version + 1;
+                .is_err()
+            {
+                tracing::error!(
+                    batch_start_version = batch_start_version,
+                    "Failed to update file store metadata. Retrying."
+                );
+                std::thread::sleep(std::time::Duration::from_millis(500));
+                METADATA_UPLOAD_FAILURE_COUNT.inc();
+            }
+            info!(
+                end_version = batch_start_version,
+                duration_in_secs = start_time.elapsed().as_secs_f64(),
+                service_type = SERVICE_TYPE,
+                "{}",
+                IndexerGrpcStep::FilestoreUpdateMetadata.get_label()
+            );
+            LATEST_PROCESSED_VERSION
+                .with_label_values(&[
+                    SERVICE_TYPE,
+                    IndexerGrpcStep::FilestoreUpdateMetadata.get_step(),
+                    IndexerGrpcStep::FilestoreUpdateMetadata.get_label(),
+                ])
+                .set(batch_start_version as i64);
+            DURATION_IN_SECS
+                .with_label_values(&[
+                    SERVICE_TYPE,
+                    IndexerGrpcStep::FilestoreUpdateMetadata.get_step(),
+                    IndexerGrpcStep::FilestoreUpdateMetadata.get_label(),
+                ])
+                .set(start_time.elapsed().as_secs_f64());
+            let size = last_version - first_version + 1;
+            PROCESSED_VERSIONS_COUNT.inc_by(size);
```

Comment:
you should use the `NUM_TRANSACTIONS_COUNT` metric over this one which is old.

Reply:
I guess we're doing both rn? Let's do a metrics cleanup once we confirm the ones we need.